### PR TITLE
fix(tui): support legacy app-server approvals

### PR DESCRIPTION
**Pull request by fcoury**
## Summary

Add compatibility for older app-server approval request shapes.

## Problem

Some app-server sessions still use legacy approval requests instead of the current v2 request types. The TUI already had the right approval UI, but it rejected these legacy requests as unsupported, leaving the agent blocked.

## What this change does

- accepts legacy `ExecCommandApproval` requests
- bridges them into the existing exec approval UI
- preserves shell-escaped command display when bridging legacy argv
- accepts legacy `ApplyPatchApproval` requests
- bridges them into the existing file-change approval UI
- tracks whether pending approvals came from legacy or current request shapes
- serializes approval responses back using the matching v1 or v2 wire format
- keeps pending approvals retryable until resolution RPCs succeed
- removes resolved bridged approvals from buffered, queued, replay, and modal state

## Result

Sessions using older app-server approval protocols can interoperate correctly with the TUI's existing approval flows for both command execution and patch application, including display, resolution, retry, and cleanup behavior.

### CODE REVIEW

File: codex-rs/tui_app_server/src/app.rs [hunk 1]
```diff
@@ -1360,6 +1366,32 @@ impl App {
...
+                .contains(&approval.effective_approval_id())
+        });
+        self.chat_widget
+            .remove_resolved_exec_approvals(&resolved.exec_approval_ids);
+
+        for channel in self.thread_event_channels.values() {
+            let mut store = channel.store.lock().await;
+            for approval_id in &resolved.exec_approval_ids {
+                store.note_exec_approval_resolved(approval_id);
+            }
```
[the remark this answers is not in the document]
Comment (fcoury) [reply codex-rs/tui_app_server/src/app.rs:1646]: Addressed in `5e0e879a4`. `note_app_server_request_resolved()` now also drains `active_thread_rx`, removes any queued `CommandExecutionRequestApproval` whose effective approval id matches the resolved ids, and requeues unrelated events in order. That closes the remaining race where a fast `serverRequest/resolved` could still replay a dead approval prompt from the active thread queue. I also added a regression test that enqueues an active-thread approval, resolves it, and asserts the live queue is empty afterward.

File: codex-rs/tui_app_server/src/bottom_pane/approval_overlay.rs [hunk 2]
```diff
@@ -465,6 +491,10 @@ impl BottomPaneView for ApprovalOverlay {
         self.enqueue_request(request);
         None
     }
+
+    fn remove_resolved_exec_approvals(&mut self, approval_ids: &[String]) -> bool {
+        self.remove_resolved_exec_approvals(approval_ids)
```
[the remark this answers is not in the document]
Comment (fcoury) [reply codex-rs/tui_app_server/src/bottom_pane/approval_overlay.rs]: Addressed in `ca09ea47c`. I renamed the inherent `ApprovalOverlay` helper to `remove_resolved_exec_approvals_inner` and delegated to that from the `BottomPaneView` impl, so the trait method no longer resolves back to itself. This preserves the resolved-approval cleanup behavior without the recursion/stack-overflow path.

File: codex-rs/tui_app_server/src/bottom_pane/mod.rs [hunk 3]
```diff
@@ -913,6 +913,17 @@ impl BottomPane {
         self.push_view(Box::new(modal));
     }
 
+    pub(crate) fn remove_resolved_exec_approvals(&mut self, approval_ids: &[String]) {
+        if let Some(view) = self.view_stack.last_mut()
+            && view.remove_resolved_exec_approvals(approval_ids)
+        {
```
[the remark this answers is not in the document]
Comment (fcoury) [reply codex-rs/tui_app_server/src/bottom_pane/mod.rs]: Addressed in `89efa807e`. `BottomPane::remove_resolved_exec_approvals()` now walks the full `view_stack`, removes resolved exec approvals from every view, and drops any views that become complete. I added regression coverage for the hidden-overlay case by placing a user-input modal on top of an approval overlay and verifying the stale approval cannot resurface after the top modal closes.

File: codex-rs/tui_app_server/src/app/app_server_adapter.rs [hunk 4]
```diff
@@ -283,6 +367,44 @@ impl App {
...
+        ServerRequest::CommandExecutionRequestApproval {
+            request_id: request_id.clone(),
+            params: CommandExecutionRequestApprovalParams {
+                thread_id: params.conversation_id.to_string(),
+                turn_id: String::new(),
+                item_id: params.call_id.clone(),
+                approval_id: params.approval_id.clone(),
+                reason: params.reason.clone(),
+                network_approval_context: None,
+                command: Some(params.command.join(" ")),
```
[the remark this answers is not in the document]
Comment (fcoury) [reply codex-rs/tui_app_server/src/app/app_server_adapter.rs]: Addressed in `89efa807e`. The legacy exec-approval bridge now uses the existing argv escaping helper instead of `join(" ")`, so the bridged command string preserves quoting for spaced or shell-special arguments. I also tightened the bridge regression test to cover a quoted argv case.

File: codex-rs/tui_app_server/src/bottom_pane/mod.rs [hunk 5]
```diff
@@ -913,6 +913,17 @@ impl BottomPane {
         self.push_view(Box::new(modal));
     }
 
+    pub(crate) fn remove_resolved_exec_approvals(&mut self, approval_ids: &[String]) {
+        let mut changed = false;
+        self.view_stack.retain_mut(|view| {
+            changed |= view.remove_resolved_exec_approvals(approval_ids);
+            !view.is_complete()
+        });
+        if changed {
```
[the remark this answers is not in the document]
Comment (fcoury) [reply codex-rs/tui_app_server/src/bottom_pane/mod.rs:926]: Addressed in 20358f7353a03aabf47367023c350f5d41bf4bed. `BottomPane::remove_resolved_exec_approvals()` now detects when it removed the active modal view and only then runs the normal teardown path, which resumes the paused status timer. I also added a regression test that exercises active exec-approval removal while a task is running.

File: codex-rs/tui_app_server/src/app/app_server_requests.rs [hunk 6]
```diff
@@ -99,13 +126,20 @@ impl PendingAppServerRequests {
...
-                    request_id: request_id.clone(),
-                    message:
-                        "Legacy command approval requests are not available in app-server TUI yet."
-                            .to_string(),
-                })
+            ServerRequest::ExecCommandApproval { request_id, params } => {
+                let approval_id = params
+                    .approval_id
+                    .clone()
+                    .unwrap_or_else(|| params.call_id.clone());
```
[the remark this answers is not in the document]
Comment (fcoury) [reply codex-rs/tui_app_server/src/app/app_server_requests.rs:156]: Addressed in `af568a7c6`. `handle_server_request_event()` now treats every legacy `ExecCommandApproval` through the extracted helper: remote sessions still bridge into the v2 thread-scoped approval path, while embedded/local sessions now reject the legacy request immediately as unsupported before it is recorded in `PendingAppServerRequests`. That keeps unsupported local legacy approvals from being dropped threadless and stalling the turn without a reject RPC.

File: codex-rs/tui_app_server/src/app.rs [hunk 7]
```diff
@@ -1577,6 +1577,98 @@ impl App {
...
+        }
+
+        if disconnected {
+            self.clear_active_thread().await;
+            return;
+        }
+
+        for event in retained_events {
+            if sender.send(event).await.is_err() {
```
[the remark this answers is not in the document]
Comment (fcoury) [reply codex-rs/tui_app_server/src/app.rs]: Addressed in `af568a7c6`. `remove_resolved_exec_approvals_from_active_queue()` now restores `active_thread_rx` before reinserting retained events, uses `try_send` for the fast path, and falls back to a spawned sequential refill only after the receiver is back in place. That removes the awaitable `send()` while the receiver is detached, so resolved-approval cleanup can no longer deadlock on channel backpressure.

File: codex-rs/tui_app_server/src/app.rs [hunk 8]
```diff
@@ -1577,6 +1605,144 @@ impl App {
...
+                continue;
+            };
+
+            let mut retained_events = Vec::new();
+            loop {
+                match rx.try_recv() {
+                    Ok(event) => {
+                        let ThreadBufferedEvent::Request(
+                            ServerRequest::CommandExecutionRequestApproval { params, .. },
+                        ) = &event
```
[the remark this answers is not in the document]
Comment (fcoury) [reply codex-rs/tui_app_server/src/app.rs]: Addressed in 71cb7cfa6.

Rather than track and cancel queued fallback send tasks for this deprecated legacy approval path, the TUI now drops stale `CommandExecutionRequestApproval` events at consume time if their effective approval id is no longer pending. That keeps the fix scoped to the user-visible failure mode without adding broader queue-cancellation state.

I also added regression coverage for the stale consume path.

File: codex-rs/tui_app_server/src/app.rs [hunk 9]
```diff
@@ -1577,6 +1577,120 @@ impl App {
...
+
+        self.refresh_pending_thread_approvals().await;
+    }
+
+    async fn remove_resolved_exec_approvals_from_active_queue(&mut self, approval_ids: &[String]) {
+        let Some(thread_id) = self.active_thread_id else {
+            return;
+        };
+        let Some(mut rx) = self.active_thread_rx.take() else {
+            return;
```
[the remark this answers is not in the document]
Comment (fcoury) [reply codex-rs/tui_app_server/src/app.rs:1659]: Addressed by the current branch.

This was first fixed in `c34865c54`, where `note_app_server_request_resolved()` started scrubbing inactive `ThreadEventChannel.receiver` queues via `remove_resolved_exec_approvals_from_inactive_queues(...)` so switched-away threads no longer retain resolved `CommandExecutionRequestApproval` events.

There is also a later consume-time guard in `71cb7cfa6` that drops any stale exec approval whose effective approval id is no longer pending, which closes the remaining delayed-send race as well. Together those changes keep inactive queues from resurfacing resolved approval prompts.

File: codex-rs/tui_app_server/src/app.rs [hunk 10]
```diff
@@ -4790,6 +4958,19 @@ impl App {
...
+        let ServerRequest::CommandExecutionRequestApproval { params, .. } = request else {
+            return false;
+        };
+        let approval_id = params
+            .approval_id
+            .clone()
+            .unwrap_or_else(|| params.item_id.clone());
+        !self
+            .pending_app_server_requests
+            .has_pending_exec_approval(&approval_id)
```
[the remark this answers is not in the document]
Comment (fcoury) [reply codex-rs/tui_app_server/src/app.rs:4989]: Addressed in `baebae315`.

I changed app-server request resolution to a prepare/commit flow instead of removing the pending entry up front. `try_resolve_app_server_request()` now prepares the response payload, calls `resolve_server_request(...)`, and only commits the pending-entry removal after the RPC succeeds.

That keeps exec approvals retryable when the resolution RPC fails, so replay and the consume-time stale-request guard no longer suppress a prompt that still has pending work behind it. I also added regression coverage in `app_server_requests.rs` for the pending-until-commit behavior.

File: codex-rs/tui_app_server/src/app.rs [hunk 11]
```diff
@@ -1577,6 +1605,140 @@ impl App {
...
+        if disconnected {
+            self.clear_active_thread().await;
+            return;
+        }
+
+        self.active_thread_rx = Some(rx);
+
+        if retained_events.is_empty() {
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (fcoury) [reply codex-rs/tui_app_server/src/app.rs:1705]: Thanks. I looked at this again, but I am going to leave it as-is.

This requeue path is intentionally using the same ordering model as the rest of the thread-channel pipeline: once a channel hits backpressure, `enqueue_thread_request()` and the sibling enqueue helpers already fall back to spawned `sender.send(...)` tasks, so strict total ordering versus concurrent arrivals is not a guarantee the app provides today.

For this PR, the important correctness issue is preventing resolved legacy exec approvals from resurfacing. That is already covered by the queue scrubbing plus the later consume-time stale-request guard. Tightening only this scrub path into a stronger ordering guarantee would add disproportionate complexity for a deprecated compatibility flow without making the surrounding queue semantics uniformly ordered.

File: codex-rs/tui_app_server/src/app/app_server_adapter.rs [hunk 12]
```diff
@@ -283,6 +382,44 @@ impl App {
...
+        ServerRequest::CommandExecutionRequestApproval {
+            request_id: request_id.clone(),
+            params: CommandExecutionRequestApprovalParams {
+                thread_id: params.conversation_id.to_string(),
+                turn_id: String::new(),
+                item_id: params.call_id.clone(),
+                approval_id: params.approval_id.clone(),
+                reason: params.reason.clone(),
+                network_approval_context: None,
+                command: Some(escape_command(&params.command)),
```
[the remark this answers is not in the document]
Comment (fcoury) [reply codex-rs/tui_app_server/src/app/app_server_adapter.rs]: Addressed in `bdc1274be`.

I fixed this at the adapter boundary instead of in the legacy bridge itself. `exec_approval_request_from_params(...)` now uses `shlex::split` to reconstruct argv from `commandExecution/requestApproval.command`, with a single-string fallback only if parsing fails.

That keeps bridged legacy approvals from being displayed as a pre-escaped one-token command and makes the app-server TUI path handle shell-text commands more cleanly in general. I also added regression coverage for the quoted-command case.
